include/gunrock/algorithms/async/bfs.cuh

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <gunrock/algorithms/algorithms.hxx>
+#include <gunrock/algorithms/async/async_enactor.hxx>
+
+namespace gunrock {
+namespace async {
+namespace bfs {
+
+// !! Identical to synchronous `essentials`
+template <typename vertex_t>
+struct param_t {
+  vertex_t single_source;
+  param_t(vertex_t _single_source) : single_source(_single_source) {}
+};
+
+// !! Identical to synchronous `essentials`
+template <typename edge_t>
+struct result_t {
+  edge_t* depth;
+  result_t(edge_t* _depth) : depth(_depth) {}
+};
+
+// !! Identical to synchronous `essentials`
+template <typename graph_t, typename param_type, typename result_type>
+struct problem_t : gunrock::problem_t<graph_t> {
+
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t   = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+  
+  param_type param;
+  result_type result;
+
+  problem_t(graph_t& G,
+            param_type& _param,
+            result_type& _result,
+            std::shared_ptr<cuda::multi_context_t> _context)
+      : gunrock::problem_t<graph_t>(G, _context),
+        param(_param),
+        result(_result) {}
+  
+  void init() override {
+    // noop
+  }
+  
+  void reset() override {
+    auto g = this->get_graph();
+    auto n_vertices = g.get_number_of_vertices();
+    
+    auto context = this->get_single_context();
+    auto policy  = context->execution_policy();
+    
+    auto single_source = param.single_source;
+    auto d_depth       = thrust::device_pointer_cast(this->result.depth);
+    thrust::fill_n(
+      policy, d_depth, n_vertices, std::numeric_limits<vertex_t>::max());
+    thrust::fill_n(
+      policy, d_depth + single_source, 1, 0);
+  }
+};
+
+// --
+// Enactor
+
+// !! This is annoying ... ideally we'd be able to initialize the frontier
+//    w/ a lambda or thrust call.  But not sure how to get that to work.
+template<typename queue_t, typename val_t>
+__global__ void _push_one(queue_t q, val_t val) {
+    q.push(val);
+}
+
+template <typename problem_t>
+struct enactor_t : async_enactor_t<problem_t> {
+  
```

Comment:
Inherit from `async_enactor_t`, but otherwise used much the same as `enactor_t`.  User only has to see / write application-specific code in this file.

Reply:
I really, really like this. We just inherit the async enactor and ideally, everything else remains the same.